File: Back office/classes/Point.php
<?php

/**
 * Copyright 2012-2013 Mairie de Paris
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
**/

defined('EXEC') or die('Access forbidden');

class Point {

    private $id = 0;
    private $lat = 0.;
    private $lon = 0.;
    private $data = array();
    private $time;
    private $sourced = 0;
    private $count = 0;

    public function __construct(stdClass $row) {
        $this->id = intval($row->id);
        $this->sourced = intval($row->sourced);
        $this->time = intval($row->update);
        $this->lat = doubleval($row->lat);
        $this->lon = doubleval($row->lon);
        $this->data = unserialize($row->data);
        $this->count = intval($row->count);
    }

    public static function create($dataset, $lat, $lon, array $data) {
        $query = 'INSERT INTO `points`
                (`dataset`, `sourced`, `lat`, `lon`, `update`, `data`)
                VALUES
                (' . intval($dataset) . ',
                 1,
                 ' . floatval($lat) . ',
                 ' . floatval($lon) . ',
                 ' . intval(time()) . ',
                 ' . DBO::escape(serialize($data)) . ');';
        return (DBO::query($query));
    }

    public function plusOne(array $data) {
        DBO::query('UPDATE `points`
                SET `count` = `count` + 1,
                `data` = ' . DBO::escape(serialize($data)) . ',
                `update` = ' . time() . '
                WHERE `id` = ' . intval($this->id));
    }

    public static function getFromParsor($parsor, $lastmod = 0) {
        $list = array();

        $res = DBO::query('SELECT *
                    FROM `points`
                    WHERE `dataset` = ' . $parsor . '
                    AND `update` >= ' . intval($lastmod));

        while (($row = $res->fetch_object())) {
            $list[] = new Point($row);
        }
        return ($list);
    }

    public function getID() {
        return ($this->id);
    }

    public function getData() {
        return ($this->data);
    }

    public function getAData($data) {
        if (isset($this->data[$data])) {
            return ($this->data[$data]);
        } else {
            return (null);
        }
    }

    public static function getFromPos($dataset, $lat, $lon) {
        $query = 'SELECT * FROM `points` WHERE `dataset` = ' . intval($dataset);
        $query .= ' AND ABS(`lat` - ' . floatval($lat) . ') <= 0.00001';
        $query .= ' AND ABS(`lon` - ' . floatval($lon) . ') <= 0.00001';
        $res = DBO::query($query);
        if (($row = $res->fetch_object())) { //Already exist
            return (new Point($row));
        } else {
            return (null);
        }
    }

    public static function getFromID($id) {
        $query = 'SELECT * FROM `points` WHERE `id` = ' . intval($id);

        $res = DBO::query($query);
        if (($row = $res->fetch_object())) { //Already exist
            return (new Point($row));
        } else {
            return (null);
        }
    }

    public function getLat() {
        return ($this->lat);
    }

    public function getLon() {
        return ($this->lon);
    }

    public function getCount() {
        return ($this->count + $this->sourced);
    }

    public function getLastUpdate() {
        return ($this->time);
    }

}

File: Android app/src/com/containers/Point.java
/**
 * Copyright 2012-2013 Mairie de Paris
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at

 * http://www.apache.org/licenses/LICENSE-2.0

 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
**/

package com.containers;

import java.util.ArrayList;
import java.util.Calendar;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.database.Cursor;

import com.utils.DataBase;
import com.utils.SqliteRequestPoints;

public class Point {
	
	private ArrayList<Spec> mList;
	private String mStringList;
	private int mTimestamp;
	private int mId;
	private TypePoint mType;
	private int mLattitude;
	private int mLongitude;
	private String mImgUrl;

	/**
	 * Constructor used when adding a new point to the database
	 * 
	 * @param point
	 * @throws JSONException
	 */
	public Point(JSONObject point) throws JSONException {
		mStringList = point.getString("spec");
		mTimestamp = point.getInt("timestamp");
		mId = point.getInt("id");
		mType = new TypePoint(point.getInt("id_type"));
		mLattitude = (int) (point.getDouble("lat") * 1000000);
		mLongitude = (int) (point.getDouble("long") * 1000000);
		mImgUrl = point.getString("url_img");
	}

	/**
	 * Used to create a new point to send to the server
	 * 
	 * @param spec list of the specs
	 * @param id_type type
	 * @param lat lattitude
	 * @param lon longitude
	 * @return the points in a JSON format
	 */
	public static String getJsonPoint(JSONArray spec, int id_type, double lat,
			double lon) {
		JSONObject jo = new JSONObject();

		try {
			jo.put("spec", spec);
			jo.put("timestamp", (int) (Calendar.getInstance().getTime()
					.getTime() / 1000));
			jo.put("id", 0);
			jo.put("id_type", id_type);
			jo.put("lat", lat);
			jo.put("long", lon);
			return jo.toString();
		} catch (JSONException e) {
			e.printStackTrace();
		}
		return (null);
	}

	/**
	 * Retrieve a point from the database
	 * 
	 * @param type
	 */
	public static ArrayList<Point> getPointsFromType(int type, Context context) {
		SqliteRequestPoints bdd = new SqliteRequestPoints(context);
		ArrayList<Point> points = bdd.getPointsFromType(type, context);
		bdd.close();
		return points;
	}

	public static Point getPointFromId(Integer id, Context context) {
		SqliteRequestPoints bdd = new SqliteRequestPoints(context);
		Point point = bdd.getPointFromId(id, context);
		bdd.close();
		return point;
	}

	public Point(Cursor c, Context context) throws JSONException {

		String specs = c.getString(c.getColumnIndex(DataBase.TABLE_POINT_SPECS));
		JSONArray JSONSpecList = new JSONArray(specs);
		int size = JSONSpecList.length();
		mList = new ArrayList<Spec>();
		for (int i = 0; i < size; i++) {
			mList.add(new Spec((JSONSpecList.getJSONObject(i)), context));
		}
		mTimestamp = c.getInt(c.getColumnIndex(DataBase.TABLE_POINT_TIMESTAMP));
		mId = c.getInt(c.getColumnIndex(DataBase.TABLE_POINT_ID));
		mType = TypePoint.getTypeFromId(
				c.getInt(c.getColumnIndex(DataBase.TABLE_POINT_TYPE)), context);
		mLattitude = c.getInt(c.getColumnIndex(DataBase.TABLE_POINT_LATTITUDE));
		mLongitude = c.getInt(c.getColumnIndex(DataBase.TABLE_POINT_LONGITUDE));
		mImgUrl = c.getString(c.getColumnIndex(DataBase.TABLE_POINT_URL_IMG));

	}

	public int getTimestamp() {
		return mTimestamp;
	}

	public int getId() {
		return mId;
	}

	public TypePoint getType() {
		return mType;
	}

	public int getLattitude() {
		return mLattitude;
	}

	public int getLongitude() {
		return mLongitude;
	}

	public ArrayList<Spec> getList() {
		return mList;
	}

	public String getStringList() {
		return mStringList;
	}

	public String getImgUrl() {
		return mImgUrl;
	}

	@Override
	public String toString() {
		StringBuilder sb = new StringBuilder().append(" POINT {").append("id=")
				.append(this.mId).append(" latt=").append(this.mLattitude)
				.append(" long=").append(this.mLongitude).append(" timestamp=")
				.append(this.mTimestamp).append(" type=")
				.append(this.getType()).append(" list=").append(this.getList())
				.append(" mImgUrl=").append(this.getImgUrl()).append("}");

		return sb.toString();
	}

	public static Point getStandardPointFromType(int type, Context context) {
		SqliteRequestPoints bdd = new SqliteRequestPoints(context);
		Point point = bdd.getStandardPointFromType(type, context);
		bdd.close();
		return point;
	}

}
